### ISSUE TITLE
Concurrency problem in parser/processor initialisation

### ISSUE THREAD

**Issue by summon128** [uniVocity/univocity-parsers#186]
I have recently started using univocity parsers and stumbled a potential concurrency problem using the CsvParser with a BeanProcessor in version 2.5.3 of univocity-parsers.

The `CsvParser.beginParsing()` method causes a ConcurrentCharLoader to be created, that does some initialisation in a Thread it creates. 
As a result the follow-up call to `processor.processStarted()` will run in parallel and may execute before the Thread of ConcurrentCharLoader finishes. The BeanProcessor will then try to read the header row of the file.
If at that point the stream is still in use by the concurrent Thread then it will fail to determine the header, causing the initialisation as a whole to become corrupt.

**Comment by jbax** [comment]
Thanks but I'm unable to reproduce this. Can you provide a test case?

I tried with this (to make the input slow) but it always works:

```
     	@Test
	public void testConcurrentInitialization() throws Exception {
		CsvParserSettings settings = new CsvParserSettings();
		final boolean[] processorRan = new boolean[]{false};
		settings.setProcessor(new AbstractProcessor<ParsingContext>() {
			@Override
			public void processStarted(ParsingContext context) {
				processorRan[0] = true;
				assertEquals(context.headers(), new String[]{"a", "b"});
			}
		});
		settings.setHeaderExtractionEnabled(true);

		CsvParser parser = new CsvParser(settings);
		parser.beginParsing(new StringReader("a,b\n1,2"){
			@Override
			public int read(char[] cbuf, int off, int len) throws IOException {
				try {
					Thread.sleep(1000);
				} catch(Exception e){

				}
				return super.read(cbuf, off, len);
			}

			@Override
			public void close() {
				try {
					Thread.sleep(1000);
				} catch(Exception e){

				}
				super.close();
			}
		});


		assertTrue(processorRan[0]);
	}
```

**Comment by summon128** [comment]
Hello jbax,

thanks for your fast response and the effort of setting up a test case.

I did some more debugging based on the test case and was able to further narrow down the problem:

For the test case try replacing

`parser.beginParsing(new StringReader("a,b\n1,2"){`
with 
`parser.beginParsing(new ByteArrayInputStream("a,b\n1,2".getBytes()) {`

This will cause the parser to delegate Reader creation to ArgumentUtils, where the InputStream is wrapped in a BomInput.
As a result the first bucket received by the ConcurrentCharLoader will be filled with only 1 char at first.
At this point the main thread will no longer be blocked and will run concurrently with the _Character reading thread_ which keeps on reading more chars from the input.

If the main thread is faster, the header will return as only one column with one char.

Since `BomInput.read()` does not delegate to the underlying stream the second time it is called, overriding the read `InputStream.read()` method will not help in reproducing.
The only thing I could come up with is setting a breakpoint in `ConcurrentCharLoader:84` and resuming once.

**Comment by jbax** [comment]
Thanks!, that was painful! I could not reproduce the error with the breakpoint you suggested but after a lot of digging I was able to make this reproducible.

With the test case you mentioned, when using a `BomInput` the concurrent reader is loading a bucket of characters of size 1 (with `a`) and if the main thread does not consume that character before the next character bucket is read (which triggers the unwrapping from the BomInput), that first 1-sized buffer (with `a`) was being discarded. When the main thread finally decided to consume a character, it would find `,` and miss the first char. That is likely to happen if you are debugging something in the main thread.

You can reproduce this consistently by adding a `Thread.sleep` in the last line of `ConcurrentCharInputReader.setReader`. I made a few adjustments be able to write a test case with that.

The fix is just making sure that all buckets are consumed before unwrapping from the `BomInput` (see commit https://github.com/uniVocity/univocity-parsers/commit/d04d42b60f88115299c858acdc781aaccb1b20e3).

I've released a 2.5.5-SNAPSHOT build that includes the adjustment. Hopefully this nails everything.

Thanks a lot for reporting this error and for using our parsers!

**Comment by jbax** [comment]
Version 2.5.5 released and problem solved. Thank you for using our parsers!
